flink-streaming-connectors/flink-connector-activemq/src/main/java/org/apache/flink/streaming/connectors/activemq/AMQSource.java

```diff
@@ -0,0 +1,258 @@
+ * an array of bytes into an instance of the output type. If an incoming
+ * message is not a message with an array of bytes, this message is ignored
+ * and warning message is logged.
+ *
+ * @param <OUT> type of output messages
+ */
+public class AMQSource<OUT> extends MessageAcknowledgingSourceBase<OUT, String>
+	implements ResultTypeQueryable<OUT> {
+
+	private static final Logger LOG = LoggerFactory.getLogger(AMQSource.class);
+
+	private final ActiveMQConnectionFactory connectionFactory;
+	private final String queueName;
+	private final DeserializationSchema<OUT> deserializationSchema;
+	private boolean logFailuresOnly = false;
+	private RunningChecker runningChecker;
+	private transient Connection connection;
+	private transient Session session;
+	private transient MessageConsumer consumer;
+	private boolean autoAck;
+	private HashMap<String, Message> unaknowledgedMessages = new HashMap<>();
+
+	/**
+	 * Create AMQSource.
+	 *
+	 * @param connectionFactory factory that will be used to create a connection with ActiveMQ
+	 * @param queueName name of an ActiveMQ queue to read from
+	 * @param deserializationSchema schema to deserialize incoming messages
+	 */
+	public AMQSource(ActiveMQConnectionFactory connectionFactory, String queueName, DeserializationSchema<OUT> deserializationSchema) {
+		this(connectionFactory, queueName, deserializationSchema, new RunningCheckerImpl());
+	}
+
+	/**
+	 * Create AMQSource.
+	 *
+	 * @param connectionFactory factory that will be used to create a connection with ActiveMQ
+	 * @param queueName name of an ActiveMQ queue to read from
+	 * @param deserializationSchema schema to deserialize incoming messages
+	 * @param runningChecker running checker that is used to decide if the source is still running
+	 */
+	AMQSource(ActiveMQConnectionFactory connectionFactory, String queueName, DeserializationSchema<OUT> deserializationSchema, RunningChecker runningChecker) {
+		super(String.class);
+		this.connectionFactory = connectionFactory;
+		this.queueName = queueName;
+		this.deserializationSchema = deserializationSchema;
+		this.runningChecker = runningChecker;
+	}
+
+	/**
+	 * Defines whether the producer should fail on errors, or only log them.
+	 * If this is set to true, then exceptions will be only logged, if set to false,
+	 * exceptions will be eventually thrown and cause the streaming program to
+	 * fail (and enter recovery).
+	 *
+	 * @param logFailuresOnly The flag to indicate logging-only on exceptions.
+	 */
+	public void setLogFailuresOnly(boolean logFailuresOnly) {
+		this.logFailuresOnly = logFailuresOnly;
+	}
+
+	@Override
+	public void open(Configuration config) throws Exception {
+		super.open(config);
+		// Create a Connection
+		connection = connectionFactory.createConnection();
+		connection.start();
+
+		connection.setExceptionListener(new ExceptionListener() {
+			@Override
+			public void onException(JMSException e) {
+				LOG.error("Received ActiveMQ exception", e);
```

Comment:
I don't think I agree with this. **close**/**invoke**/**process** methods may throw get exceptions because of transient errors or because of issues we do not necessarily care about, while an exception in the **open** method may be caused by a persistent issue, such as bad configuration.
Also, if open method fails we won't be able to use the sink/source, so it will be pointless to proceed.